decisions/0004-sync-routes-from-k8s-to-cc.md

```diff
@@ -0,0 +1,113 @@
+1. [k8s.io/client-go/cache](https://godoc.org/k8s.io/client-go/tools/cache) provides a mechanism for clients to build and maintain local caches of kubernetes state
+1. [kubebuilder](https://book.kubebuilder.io/) Provides, amongst other things, an SDK for building reconciliation loops on top of watches and client side caching
+
+### Problems and Solutions
+
+To illustrate the necessity of these sync loops, it helps to think about how they improve resiliency in 2 cases of temporary network partition from the perspective of the CF API: 
+1. Problem: Updates to CCDB succeed ✅, but updates to the Route CRD fail ❌
+1. Problem: Updates to the Route CRD succeed ✅, but updates to CCDB fail ❌
+
+To handle these 2 failure modes, several approaches can be taken. 
+
+##### Mimic historical sync loops:
+1. Problem: CCDB ✅ & Route CRD ❌
+   1. Solution: Raise the error to the user, but allow the CCDB transaction to commit. A bulk sync loop will detect that the Route CRD is "missing" and create it later.
+1. Problem: Route CRD ✅ & CCDB ❌
+   1. Solution A: Raise the error to the user, and don't do the Route CRD write until CCDB changes are committed.
+   1. Solution B: A bulk sync loop will detect that the Route CRD is "extraneous" and delete it later.
+
+##### Forgo a bulk sync and attempt to bootstrap consistency with k8s off CCDB transactions:
+1. Problem: CCDB ✅ & Route CRD ❌
+   1. Solution: Raise the error to the user, abort the CCDB transaction.
+1. Problem: Route CRD ✅ & CCDB ❌
+   1. Solution: Roll back changes to the Route CRD... (but what if that still fails due to network partition?!)
+
+##### Sync from k8s to CCDB
+1. Problem: CCDB ✅ & Route CRD ❌
+   1. Solution: Raise the error to the user, abort the CCDB transaction.
+1. Problem: Route CRD ✅ & CCDB ❌
+   1. Solution: Raise the error to the user, but leave the Route CRD. The "bulk sync loop" will not consider the route "reconciled" until it is correct in CCDB.
+
+### Viable Long Term Implementations & Tradeoffs
+
+##### CCDB->k8s Sync in Ruby
+1. Mimics ProcessesSync for efficient access to CCDB
+1. Run it in an independent process from clock_global to avoid BOSH entanglement
+1. Unclear what scale limitations we might encounter if we regularly download
+   all k8s Routes.
+
+##### CCDB->k8s Sync in Go
+1. Probably a new, independent go program
+1. Can take advantage of go-client cache & reflector to avoid repeatedly
+   downloading all k8s state.
+   1. It's unclear, however, that kubebuilder could help
+   1. What does it mean to "reconcile" a change to a single Route CRD if CCDB is authoritative?
+1. Would need to page through CF API to build picture of CF state.
+
+##### Sync from k8s->CCDB with kubebuilder controller and reloads
+1. Probably a RoutesController akin to our existing kpack BuildsController
```

Comment:
Minor nit, but gotta be real careful with the naming here or its going to get confusing with the `cf-k8s-networking` [routecontroller](https://github.com/cloudfoundry/cf-k8s-networking/tree/develop/routecontroller) already being a thing. Maybe `cf-api-routes-controller` or `cf-api-sync-*` or something? If there's going to be a bunch of these potentially it's worth thinking about. 🤔

Reply:
the deployment will be called `cf-api-controllers` under our new kubebuilder-centric naming scheme, and it'd share a process with the BuildsController, hence the camel case.